File: clients/zig/src/transport.zig
//! How a request reaches the API, and what a server on the other end is not allowed to cost.
//!
//! The transport answers the status and the bytes and knows nothing of what the API declares:
//! reading those bytes is the generated half's job, and deciding whether to send them again is the
//! client's. That is what lets one HTTP implementation serve both the hand-written event path and
//! every generated method — a generated group calls whatever `runtime.Transport` it was built on,
//! and this is the one this package ships.
//!
//! The exchange is written here rather than taken from `std.http.Client`, and that is the point: a
//! client that reads a head into a buffer it sized for itself decides how much a broken or hostile
//! server may spend on a caller's behalf, and the shared conformance corpus names four separate
//! ceilings that decision has to honour. Every one of them is applied on the line that crosses it —
//! how many header lines an answer may carry, how long one of them may be, how large the whole head
//! may come to, and how many bytes of body are read off the socket — and each is a real allocation
//! bound rather than a rule somebody remembered: the buffers are sized from the bounds, so a head
//! this client will not accept is one it cannot hold.
//!
//! Every read is given what is left of the attempt's budget, worked out again before each one, so a
//! server answering one byte just under the timeout cannot hold a caller for as long as it likes.
//! Connecting is the one step that is not: `std.Io.Threaded` in Zig 0.16.0 refuses a
//! `ConnectOptions.timeout` outright — `@panic("TODO implement netConnectIpPosix with timeout")` —
//! so that step is bounded by the kernel's own connect timeout until the standard library implements
//! it. It is stated here rather than left for somebody to find.

const std = @import("std");

const runtime = @import("runtime.zig");

const net = std.Io.net;

/// Which of the corpus's causes a failure that produced no answer is.
///
/// The three are told apart because only one of them could end differently. A request that got no
/// answer — a connection refused or reset, an attempt out of time, a body that stopped mid-way —
/// says nothing about whether the API acted on it, which is exactly why a send carries an identifier
/// the client chose itself, and why repeating it is safe and worth doing. An answer that crossed a
/// ceiling this client set for itself draws the same answer the second time, and reading it again
/// four times over costs the caller four times as much for the same failure. A URL nothing can be
/// sent to was never sent at all, and a repetition builds the same unusable request, turning a
/// misconfiguration into a message that accuses the network.
///
/// The names are the ones the shared conformance corpus gives them, so the verdict a client applies
/// and the verdict that corpus writes down are the same words.
pub const Cause = enum {
    no_answer,
    answer_above_a_bound,
    unusable_api_url,

    /// Whether repeating a request that met this could end differently.
    pub fn retryable(self: Cause) bool {
        return self == .no_answer;
    }

    /// What the corpus calls this cause.
    pub fn name(self: Cause) []const u8 {
        return @tagName(self);
    }
};

/// A request the API never answered.
pub const TransportError = error{
    /// The API was reached for and answered nothing this client could read to its end.
    NoAnswer,
    /// The API answered, and what it answered crossed a ceiling this client set for itself.
    AnswerAboveABound,
    /// There is nowhere to send the request, so nothing was sent.
    UnusableApiUrl,
};

/// Which cause an error of this transport is.
pub fn causeOf(failed: TransportError) Cause {
    return switch (failed) {
        error.NoAnswer => .no_answer,
        error.AnswerAboveABound => .answer_above_a_bound,
        error.UnusableApiUrl => .unusable_api_url,
    };
}

/// Longest one attempt at reaching the API is given before it is abandoned, in milliseconds.
///
/// Ten seconds is far above what ingesting an event takes when the API is healthy, and short enough
/// that a stuck connection does not hold a caller for a noticeable time.
pub const default_request_timeout_ms: u64 = 10_000;

/// Largest response body read off a socket, in bytes.
pub const default_max_response_bytes: usize = 8 * 1024 * 1024;

/// How many header lines an answer may carry before it is refused. Sixty-four is well above what the
/// API sends.
pub const default_max_response_headers: usize = 64;

/// Longest one header line may be, name and value together, in bytes.
pub const default_max_header_bytes: usize = 64 * 1024;

/// Largest whole head an answer may carry, every line counted together, in bytes.
///
/// This is the one that bounds what a head costs. A line count and a size per line multiply:
/// sixty-four lines of sixty-four kilobytes each is four megabytes of head, and both of the bounds
/// above admit it. They earn their place by refusing early, on the line that crosses them rather
/// than at the end of the head; this one sets the ceiling.
///
/// Sixteen kilobytes is what Node enforces by default, and matching it is the point: a lower ceiling
/// would refuse heads another target accepts, and a higher one would not bind there at all, leaving
/// each language a different effective limit.
pub const default_max_head_bytes: usize = 16 * 1024;

/// What a request body says it carries, and what an answer is asked for in.
pub const json_media_type = "application/json";

/// The ceilings one exchange is held to.
pub const Bounds = struct {
    request_timeout_ms: u64 = default_request_timeout_ms,
    max_response_bytes: usize = default_max_response_bytes,
    max_response_headers: usize = default_max_response_headers,
    max_header_bytes: usize = default_max_header_bytes,
    max_head_bytes: usize = default_max_head_bytes,
};

/// Where this transport reaches, read out of a URL.
pub const Reached = struct {
    secure: bool,
    host: []const u8,
    port: u16,
    target: []const u8,

    /// What this transport reaches, refusing the text it cannot send anything to.
    ///
    /// Written here rather than taken from `std.Uri`, which reads a table of pieces out of almost
    /// any text: what a transport needs is a refusal for a URL nothing can be sent to, and that is
    /// one of the three causes the corpus classifies.
    pub fn parse(url: []const u8) TransportError!Reached {
        const at = std.mem.indexOf(u8, url, "://") orelse return error.UnusableApiUrl;
        const scheme = url[0..at];
        const rest = url[at + 3 ..];

        const secure = if (std.ascii.eqlIgnoreCase(scheme, "https"))
            true
        else if (std.ascii.eqlIgnoreCase(scheme, "http"))
            false
        else
            return error.UnusableApiUrl;

        const ends = std.mem.indexOfAny(u8, rest, "/?#") orelse rest.len;
        var authority = rest[0..ends];
        const target = if (ends == rest.len) "/" else rest[ends..];

        if (std.mem.lastIndexOfScalar(u8, authority, '@')) |credential| {
            authority = authority[credential + 1 ..];
        }

        var host = authority;
        var port: u16 = if (secure) 443 else 80;
        if (std.mem.lastIndexOfScalar(u8, authority, ':')) |colon| {
            if (std.mem.indexOfScalar(u8, authority[colon..], ']') == null) {
                host = authority[0..colon];
                port = std.fmt.parseInt(u16, authority[colon + 1 ..], 10) catch
                    return error.UnusableApiUrl;
            }
        }
        host = std.mem.trim(u8, host, "[]");
        if (host.len == 0) return error.UnusableApiUrl;

        return .{ .secure = secure, .host = host, .port = port, .target = target };
    }
};

/// One exchange with the API, bounded on every axis a server controls.
pub const Transport = struct {
    io: std.Io,
    base_url: []const u8,
    token: []const u8,
    bounds: Bounds = .{},

    /// What the generated half issues its requests through.
    pub fn any(self: *Transport) runtime.Transport {
        return .{ .context = self, .requestFn = issue };
    }

    fn issue(
        context: *anyopaque,
        allocator: std.mem.Allocator,
        asked: runtime.Request,
    ) anyerror!runtime.Answer {
        const self: *Transport = @ptrCast(@alignCast(context));
        return self.request(allocator, asked);
    }

    /// What the API answered, headers included, whether or not it answered a success.
    ///
    /// Header names are lowercased and a later value wins over an earlier one under the same name,
    /// so a caller reads a header without knowing which case the server wrote it in.
    pub fn deliver(
        self: *Transport,
        allocator: std.mem.Allocator,
        asked: runtime.Request,
    ) !Delivered {
        const reached = try self.resolved(allocator, asked);
        const deadline = self.deadlineFromNow();

        var written: ?[]const u8 = null;
        if (asked.body) |body| {
            var out: std.Io.Writer.Allocating = .init(allocator);
            var stringify: std.json.Stringify = .{ .writer = &out.writer };
            try stringify.write(body);
            written = out.written();
        }

        // The address is parsed rather than resolved: what this client is pointed at is a host it
        // was configured with, and a name that has to be looked up is resolved through `Io`.
        const address = net.IpAddress.parse(reached.host, reached.port) catch
            net.IpAddress.resolve(self.io, reached.host, reached.port) catch
            return error.UnusableApiUrl;

        const stream = address.connect(self.io, .{ .mode = .stream }) catch return error.NoAnswer;
        defer stream.close(self.io);

        try self.send(stream, reached, asked.method, written);
        return self.readAnswer(allocator, stream, deadline);
    }

    /// What the API answered, which is the status and the bytes.
    pub fn request(
        self: *Transport,
        allocator: std.mem.Allocator,
        asked: runtime.Request,
    ) !runtime.Answer {
        const answered = try self.deliver(allocator, asked);
        return .{ .status = answered.status, .payload = answered.payload };
    }

    /// What one exchange answered, beside the body.
    pub const Delivered = struct {
        status: u16,
        payload: []const u8,
        /// The headers the answer carried, lowercased.
        headers: []const Header,

        pub const Header = struct { name: []const u8, value: []const u8 };

        /// What that header carried, when the answer carried it.
        pub fn get(self: Delivered, name: []const u8) ?[]const u8 {
            for (self.headers) |header| {
                if (std.mem.eql(u8, header.name, name)) return header.value;
            }
            return null;
        }
    };

    fn deadlineFromNow(self: *Transport) std.Io.Timestamp {
        const now = std.Io.Timestamp.now(self.io, .awake);
        return now.addDuration(.fromNanoseconds(
            @as(i96, @intCast(self.bounds.request_timeout_ms)) * std.time.ns_per_ms,
        ));
    }

    /// What is left of the budget one attempt was given, worked out again before every read.
    fn within(self: *Transport, deadline: std.Io.Timestamp) TransportError!std.Io.Timeout {
        const now = std.Io.Timestamp.now(self.io, .awake);
        const left = now.durationTo(deadline);
        if (left.nanoseconds <= 0) return error.NoAnswer;
        return .{ .duration = .{ .raw = left, .clock = .awake } };
    }

    /// Where the request lands: the path of the base URL, extended by the operation's own, and then
    /// the query the operation asked for.
    fn resolved(
        self: *Transport,
        allocator: std.mem.Allocator,
        asked: runtime.Request,
    ) !Reached {
        var reached = try Reached.parse(self.base_url);

        var target: std.ArrayList(u8) = .empty;
        if (asked.path.len > 0 and asked.path[0] == '/') {
            try target.appendSlice(allocator, asked.path);
        } else {
            try target.appendSlice(allocator, std.mem.trimEnd(u8, reached.target, "/"));
            try target.append(allocator, '/');
            try target.appendSlice(allocator, asked.path);
        }

        var asking = false;
        for (asked.query) |pair| {
            var buffer: [runtime.max_written_value_bytes]u8 = undefined;
            const carried = pair.value.written(&buffer) orelse continue;
            try target.append(allocator, if (asking) '&' else '?');
            asking = true;
            try target.appendSlice(allocator, try runtime.encoded(allocator, pair.name));
            try target.append(allocator, '=');
            try target.appendSlice(allocator, try runtime.encoded(allocator, carried));
        }

        reached.target = try target.toOwnedSlice(allocator);
        return reached;
    }

    fn send(
        self: *Transport,
        stream: net.Stream,
        reached: Reached,
        method: []const u8,
        written: ?[]const u8,
    ) TransportError!void {
        var buffer: [8 * 1024]u8 = undefined;
        var out = stream.writer(self.io, &buffer);
        const writer = &out.interface;

        writer.print("{s} {s} HTTP/1.1\r\n", .{ method, reached.target }) catch return error.NoAnswer;
        writer.print("Host: {s}:{d}\r\n", .{ reached.host, reached.port }) catch return error.NoAnswer;
        writer.print("Authorization: Bearer {s}\r\n", .{self.token}) catch return error.NoAnswer;
        writer.print("Accept: {s}\r\n", .{json_media_type}) catch return error.NoAnswer;
        writer.writeAll("Connection: close\r\n") catch return error.NoAnswer;
        if (written) |body| {
            writer.print("Content-Type: {s}\r\n", .{json_media_type}) catch return error.NoAnswer;
            writer.print("Content-Length: {d}\r\n", .{body.len}) catch return error.NoAnswer;
        }
        writer.writeAll("\r\n") catch return error.NoAnswer;
        if (written) |body| {
            writer.writeAll(body) catch return error.NoAnswer;
        }
        writer.flush() catch return error.NoAnswer;
    }

    /// Everything the answer is, read under every ceiling this client set for itself.
    ///
    /// The bytes arrive through `receiveTimeout`, which is the one read in the standard library that
    /// takes a deadline, and they are held in one buffer sized from the head and body bounds: the
    /// ceilings are what the allocation is, rather than something checked after the fact.
    fn readAnswer(
        self: *Transport,
        allocator: std.mem.Allocator,
        stream: net.Stream,
        deadline: std.Io.Timestamp,
    ) !Delivered {
        var held: std.ArrayList(u8) = .empty;
        var chunk: [16 * 1024]u8 = undefined;

        var status: ?u16 = null;
        var headers: std.ArrayList(Delivered.Header) = .empty;
        var head_bytes: usize = 0;
        var read_from: usize = 0;
        var body_from: ?usize = null;

        while (true) {
            // Every line of the head is settled out of what has already arrived before another read
            // is asked for, so a head above a ceiling is refused on the line that crosses it.
            while (body_from == null) {
                const line_end = std.mem.indexOfScalarPos(u8, held.items, read_from, '\n') orelse break;
                const line = std.mem.trimEnd(u8, held.items[read_from..line_end], "\r");
                read_from = line_end + 1;

                if (status == null) {
                    if (line.len > self.bounds.max_header_bytes) return error.AnswerAboveABound;
                    const answered = std.mem.indexOfScalar(u8, line, ' ') orelse return error.NoAnswer;
                    status = std.fmt.parseInt(u16, line[answered + 1 .. answered + 4], 10) catch
                        return error.NoAnswer;
                    continue;
                }
                if (line.len == 0) {
                    body_from = read_from;
                    break;
                }

                if (line.len > self.bounds.max_header_bytes) return error.AnswerAboveABound;
                if (headers.items.len == self.bounds.max_response_headers) return error.AnswerAboveABound;
                head_bytes += line.len;
                if (head_bytes > self.bounds.max_head_bytes) return error.AnswerAboveABound;

                const colon = std.mem.indexOfScalar(u8, line, ':') orelse continue;
                try headers.append(allocator, .{
                    .name = try std.ascii.allocLowerString(allocator, line[0..colon]),
                    .value = std.mem.trim(u8, line[colon + 1 ..], " \t"),
                });
            }

            if (body_from) |from| {
                if (held.items.len - from > self.bounds.max_response_bytes) return error.AnswerAboveABound;
            } else if (held.items.len > self.bounds.max_head_bytes + self.bounds.max_header_bytes) {
                return error.AnswerAboveABound;
            }

            const timeout = try self.within(deadline);
            const message = stream.socket.receiveTimeout(self.io, &chunk, timeout) catch |failed| switch (failed) {
                error.Timeout => return error.NoAnswer,
                else => return error.NoAnswer,
            };
            if (message.data.len == 0) break;
            try held.appendSlice(allocator, message.data);
        }

        const from = body_from orelse return error.NoAnswer;
        const payload = held.items[from..];
        if (payload.len > self.bounds.max_response_bytes) return error.AnswerAboveABound;

        return .{
            .status = status orelse return error.NoAnswer,
            .payload = payload,
            .headers = try headers.toOwnedSlice(allocator),
        };
    }
};

File: clients/lua/src/transport.lua
--- How a request reaches the API, and what a server on the other end is not allowed to cost.
---
--- The transport answers the status and the bytes and knows nothing of what the API declares:
--- reading those bytes is the generated half's job, and deciding whether to send them again is the
--- client's. That is what lets one HTTP implementation serve both the hand-written event path and
--- every generated method — a generated group calls whatever object it is handed, and this is the
--- one this rock ships.
---
--- The exchange is written here rather than taken from `socket.http`, and that is the whole point:
--- `socket.http` reads a head of any length and a body of any length into memory before a caller is
--- consulted, so a server that is broken or hostile decides how much a client spends. Every ceiling
--- the shared conformance corpus names is applied here instead, on the line that crosses it: how
--- many header lines an answer may carry, how long one of them may be, how large the whole head may
--- come to, and how many bytes of body are read off the socket.

local Errors = require("hook0.errors")
local Json = require("hook0.json")
local Runtime = require("hook0.runtime")
local socket = require("socket")

local Transport = {}
Transport.__index = Transport

--- A request the API never answered, and what caused that.
---
--- The three causes are told apart because only one of them could end differently. A request that
--- got no answer — a connection refused or reset, an attempt out of time, a body that stopped
--- mid-way — says nothing about whether the API acted on it, which is exactly why a send carries an
--- identifier the client chose itself, and why repeating it is safe and worth doing. An answer that
--- crossed a ceiling this client set for itself draws the same answer the second time, and reading it
--- again four times over costs the caller four times as much for the same failure. A URL nothing can
--- be sent to was never sent at all, and a repetition builds the same unusable request, turning a
--- misconfiguration into a message that accuses the network.
---
--- The names are the ones the shared conformance corpus gives them, so the verdict a client applies
--- and the verdict that corpus writes down are the same words.
Transport.TransportError = Errors.kind("TransportError", Errors.ClientError)

--- Longest one attempt at reaching the API is given before it is abandoned, in seconds.
---
--- Ten seconds is far above what ingesting an event takes when the API is healthy, and short enough
--- that a stuck connection does not hold a caller for a noticeable time.
Transport.DEFAULT_REQUEST_TIMEOUT = 10.0

--- Largest response body read off a socket, in bytes.
Transport.DEFAULT_MAX_RESPONSE_BYTES = 8 * 1024 * 1024

--- How many header lines an answer may carry before it is refused. Sixty-four is well above what the
--- API sends.
Transport.DEFAULT_MAX_RESPONSE_HEADERS = 64

--- Longest one header line may be, name and value together, in bytes.
Transport.DEFAULT_MAX_HEADER_BYTES = 64 * 1024

--- Largest whole head an answer may carry, every line counted together, in bytes.
---
--- This is the one that bounds what a head costs. A line count and a size per line multiply:
--- sixty-four lines of sixty-four kilobytes each is four megabytes of head, and both of the bounds
--- above admit it. They earn their place by refusing early, on the line that crosses them rather
--- than at the end of the head; this one sets the ceiling.
---
--- Sixteen kilobytes is what Node enforces by default, and matching it is the point: a lower ceiling
--- would refuse heads another target accepts, and a higher one would not bind there at all, leaving
--- each language a different effective limit.
Transport.DEFAULT_MAX_HEAD_BYTES = 16 * 1024

--- How many bytes of body are asked for at a time. Bounded so that a `Content-Length` a server made
--- up cannot be turned into an allocation before a single byte has arrived.
Transport.CHUNK_BYTES = 32 * 1024

--- What a request body says it carries, and what an answer is asked for in.
Transport.JSON_MEDIA_TYPE = "application/json"

--- Longest each part this client composes its `User-Agent` out of may be, in characters.
---
--- The interpreter and the operating system are described by the platform rather than by this rock,
--- so their length is not this rock's to guarantee: they are cut here so that the header cannot grow
--- with whatever the platform feels like saying. Every part is also stripped of anything the grammar
--- of the header uses as punctuation, so a platform cannot forge a shape it does not have.
local MAX_USER_AGENT_PART_CHARS = 64

--- Longest a duration this client states its retry policy in may be, in milliseconds.
---
--- The three durations of a policy are numbers a caller set, and a header is no place for whatever
--- arithmetic they lead to: about twenty-five days is already past any schedule a send could hold,
--- and cutting to it is what keeps the value an integer whatever was configured.
local MAX_STATED_MILLISECONDS = (1 << 31) - 1

--- The schemes this transport reaches, and the port each one uses when the URL names none.
local PORTS = { http = 80, https = 443 }

--- The API was reached for and answered nothing this client could read to its end.
--- @param detail string
function Transport.no_answer(detail)
  Errors.throw(Transport.TransportError, detail, { cause = "no_answer", retryable = true })
end

--- The API answered, and what it answered crossed a ceiling this client set for itself.
--- @param detail string
function Transport.answer_above_a_bound(detail)
  Errors.throw(Transport.TransportError, detail, { cause = "answer_above_a_bound", retryable = false })
end

--- There is nowhere to send the request, so nothing was sent.
--- @param detail string
function Transport.unusable_api_url(detail)
  Errors.throw(Transport.TransportError, detail, { cause = "unusable_api_url", retryable = false })
end

--- What this transport reaches, read out of a URL.
---
--- Written here rather than taken from `socket.url`, which answers a table of pieces for any text at
--- all: what a transport needs is a refusal for the text it cannot send anything to, and that is one
--- of the three causes the corpus classifies.
---
--- @param url string
--- @return table
local function reachable(url)
  if type(url) ~= "string" then
    Transport.unusable_api_url("the API URL is " .. type(url) .. ", not a URL")
  end

  local scheme, rest = url:match("^(%a[%w+.-]*)://(.*)$")
  if scheme == nil then
    Transport.unusable_api_url("`" .. Runtime.preview(url) .. "` names no scheme this transport can send a request to")
  end

  scheme = scheme:lower()
  if PORTS[scheme] == nil then
    Transport.unusable_api_url("`" .. scheme .. "` is not a scheme this transport can send a request to")
  end

  local authority = rest:match("^([^/?#]*)") or ""
  local target = rest:sub(#authority + 1)
  if authority:find("@", 1, true) then
    authority = authority:sub(authority:find("@", 1, true) + 1)
  end

  local host, port = authority:match("^%[([^%]]*)%]:?(%d*)$")
  if host == nil then
    host, port = authority:match("^([^:]*):?(%d*)$")
  end
  if host == nil or host == "" then
    Transport.unusable_api_url("`" .. Runtime.preview(url) .. "` names no host to send a request to")
  end

  return {
    scheme = scheme,
    host = host,
    port = port ~= "" and math.tointeger(tonumber(port)) or PORTS[scheme],
    target = target ~= "" and target or "/",
  }
end

--- A value as it travels in a query string.
local function encoded(text)
  return (text:gsub("[^A-Za-z0-9%-%._~]", function(character)
    return string.format("%%%02X", character:byte())
  end))
end

--- Where the request lands: the path of the base URL, extended by the operation's own, and then the
--- query the operation asked for.
local function resolved(base_url, path, query)
  local reached = reachable(base_url)
  local written = tostring(path or "")

  if written:sub(1, 1) == "/" then
    reached.target = written
  elseif written ~= "" then
    reached.target = reached.target:gsub("/*$", "") .. "/" .. written
  end

  local asked = {}
  for index = 1, #(query or {}) do
    asked[#asked + 1] = encoded(tostring(query[index][1])) .. "=" .. encoded(tostring(query[index][2]))
  end
  if #asked > 0 then
    local separator = reached.target:find("?", 1, true) and "&" or "?"
    reached.target = reached.target .. separator .. table.concat(asked, "&")
  end

  return reached
end

--- What is left of the budget one attempt was given, set on the socket before every blocking call.
---
--- A timeout set once bounds each call rather than the exchange: a server answering one byte just
--- under it, over and over, would hold a caller for as long as it liked. What is bounded here is the
--- whole attempt, so the arithmetic is done again before every read and every write.
local function within(connection, deadline)
  local left = deadline - socket.gettime()
  if left <= 0 then
    Transport.no_answer("the attempt ran out of the time it was given before the API answered")
  end
  connection:settimeout(left)
  return connection
end

--- One connection to the API, wrapped in TLS when the URL asked for it.
local function connected(reached, deadline)
  local raw, opening = socket.tcp()
  if raw == nil then
    Transport.no_answer("no socket could be opened: " .. tostring(opening))
  end

  local opened, failed = within(raw, deadline):connect(reached.host, reached.port)
  if opened == nil then
    raw:close()
    Transport.no_answer("the API at " .. reached.host .. ":" .. reached.port ..
      " was not reached: " .. tostring(failed))
  end

  if reached.scheme ~= "https" then
    return raw
  end

  local ok, ssl = pcall(require, "ssl")
  if not ok then
    raw:close()
    Transport.unusable_api_url("`https` needs luasec, which is not installed beside this rock")
  end

  local secured, wrapping = ssl.wrap(raw, { mode = "client", protocol = "any", verify = "peer", options = "all" })
  if secured == nil then
    raw:close()
    Transport.no_answer("the connection could not be secured: " .. tostring(wrapping))
  end
  secured:sni(reached.host)

  local shaken, refusing = within(secured, deadline):dohandshake()
  if shaken == nil then
    secured:close()
    Transport.no_answer("the API refused the handshake: " .. tostring(refusing))
  end
  return secured
end

--- One line of the head, refused before it is held whole when it is longer than this client reads.
local function head_line(connection, deadline, max_header_bytes)
  local line, failed, partial = within(connection, deadline):receive("*l")
  if line == nil then
    if partial ~= nil and #partial > max_header_bytes then
      Transport.answer_above_a_bound(
        "the API answered a header above the " .. max_header_bytes .. " bytes read at most")
    end
    Transport.no_answer("the API answered nothing this client could read: " .. tostring(failed))
  end
  if #line > max_header_bytes then
    Transport.answer_above_a_bound("the API answered a header above the " .. max_header_bytes .. " bytes read at most")
  end
  return line
end

--- What an answer carried beside its body, under the names a caller looks them up by.
---
--- Read line by line and refused on the line that crosses a ceiling, rather than after the head has
--- been held whole: a head this client will not accept costs it the line it was refused on.
local function carried(connection, deadline, options)
  local headers = {}
  local held = 0
  local whole = 0

  while true do
    local line = head_line(connection, deadline, options.max_header_bytes)
    if line == "" then
      return headers
    end

    held = held + 1
    if held > options.max_response_headers then
      Transport.answer_above_a_bound(
        "the API answered more than the " .. options.max_response_headers .. " header lines read at most")
    end

    whole = whole + #line
    if whole > options.max_head_bytes then
      Transport.answer_above_a_bound(
        "the API answered a head above the " .. options.max_head_bytes .. " bytes read at most")
    end

    local name, value = line:match("^([^:]+):%s*(.*)$")
    if name ~= nil then
      headers[name:lower()] = (value:gsub("%s+$", ""))
    end
  end
end

--- The body of an answer, up to what this transport agrees to hold.
---
--- Read in bounded pieces whatever the answer says its length is, so a `Content-Length` a server made
--- up costs nothing until the bytes behind it actually arrive.
local function bounded(connection, deadline, headers, max_response_bytes)
  local announced = math.tointeger(tonumber(headers["content-length"] or ""))
  local held = {}
  local size = 0

  while true do
    local wanted = Transport.CHUNK_BYTES
    if announced ~= nil then
      wanted = math.min(wanted, announced - size)
      if wanted <= 0 then
        break
      end
    end

    local read, failed, partial = within(connection, deadline):receive(wanted)
    local piece = read or partial or ""
    size = size + #piece
    if size > max_response_bytes then
      Transport.answer_above_a_bound(
        "the API answered more than the " .. max_response_bytes .. " bytes read at most")
    end
    if #piece > 0 then
      held[#held + 1] = piece
    end

    if read == nil then
      if failed == "closed" then
        break
      end
      Transport.no_answer("the API stopped answering mid-body: " .. tostring(failed))
    end
  end

  return table.concat(held)
end

--- One part of the `User-Agent`, with everything the header's own grammar uses taken out of it and
--- cut to `MAX_USER_AGENT_PART_CHARS`.
--- @param part string
--- @return string
local function clipped(part)
  local kept = part:gsub("[^\32-\126]", ""):gsub("[();]", "")
  return kept:sub(1, MAX_USER_AGENT_PART_CHARS)
end

--- What this rock says it is, composed once: none of the three parts it is built out of can change
--- while a process runs.
local composed = nil

--- Which SDK, at which version, on which runtime and operating system, is talking to the API.
---
--- The version is the one the module this file is a half of declares, rather than a second copy of
--- it here, and it is read at first use rather than at load: that module is assembled out of this
--- one, so requiring it as this file loads would be a cycle. Lua says nothing about the operating
--- system beyond the separator it writes paths with, and neither luasocket nor luasec adds to it,
--- which leaves the two families it can tell apart.
--- @return string
local function user_agent()
  if composed == nil then
    local family = package.config:sub(1, 1) == "\\" and "windows" or "posix"
    composed = "hook0-client-lua/" .. clipped(require("hook0").VERSION) ..
      " (" .. clipped(_VERSION) .. "; " .. clipped(family) .. ")"
  end
  return composed
end

--- One duration of a retry policy, in the whole milliseconds it is stated as.
---
--- What arrives here is already the duration in force, so a value no schedule could be built on has
--- been read as its default before this sees it: what is stated and what is waited are the same
--- number by construction rather than by two rules that agree today. All that is left is the
--- ceiling, which keeps a finite duration nobody meant from deciding how long the header is.
--- @param seconds number
--- @return integer
local function stated_milliseconds(seconds)
  local milliseconds = (tonumber(seconds) or 0) * 1000
  return math.floor(math.min(math.max(milliseconds, 0), MAX_STATED_MILLISECONDS) + 0.5)
end

--- The retry policy a transport was built to serve, as every request states it.
---
--- What it states is what the policy holds rather than what one send went on to do: a policy
--- allowing a single attempt still names the delays it holds, and an instance reading `attempts=1`
--- already knows none of them will be waited. It is the one client setting the API can see the
--- consequences of without being told — a burst of identical requests is a client repeating one
--- send, and nothing else on the wire tells that apart from a client in a loop.
---
--- The grammar is the one `X-Hook0-Signature` already travels under, parts joined by `,` and each
--- cut at its first `=`, so nothing here is a second shape to get wrong.
--- @param policy table
--- @return string
local function client_options(policy)
  return string.format(
    "attempts=%d,backoff=%d,ceiling=%d,budget=%d",
    policy:attempts(),
    stated_milliseconds(policy:initial_backoff_in_force()),
    stated_milliseconds(policy:max_backoff_in_force()),
    stated_milliseconds(policy:max_total_delay_in_force()))
end

--- Builds a transport pointed at one API, under one credential.
---
--- The policy a transport states is the one the client it serves was built with, and the default one
--- where nothing handed it a client. It is asked for here rather than at load, for the reason
--- `user_agent` is: the module that declares a policy is assembled out of this one.
---
--- @param base_url string where the API lives, such as https://app.hook0.com/api/v1
--- @param token string an authentication token valid for that API
--- @param options table|nil the ceilings one exchange is held to
--- @return table
function Transport.new(base_url, token, options)
  local chosen = options or {}
  return setmetatable({
    base_url = base_url,
    token = token,
    timeout = chosen.request_timeout or Transport.DEFAULT_REQUEST_TIMEOUT,
    max_response_bytes = chosen.max_response_bytes or Transport.DEFAULT_MAX_RESPONSE_BYTES,
    max_response_headers = chosen.max_response_headers or Transport.DEFAULT_MAX_RESPONSE_HEADERS,
    max_header_bytes = chosen.max_header_bytes or Transport.DEFAULT_MAX_HEADER_BYTES,
    max_head_bytes = chosen.max_head_bytes or Transport.DEFAULT_MAX_HEAD_BYTES,
    retry_policy = chosen.retry_policy or require("hook0").RetryPolicy.new(),
  }, Transport)
end

--- What the API answered, headers included, whether or not it answered a success.
---
--- Header names are lowercased and a later value wins over an earlier one under the same name, so a
--- caller reads a header without knowing which case the server wrote it in.
---
--- @param method string the HTTP method the operation is issued under
--- @param path string where the request lands, absolute or under the base URL
--- @param query table|nil the name and value pairs of the query string
--- @param body any|nil what to send as a JSON document, or nothing at all
--- @return integer, table, string the status, the headers and the body
function Transport:deliver(method, path, query, body)
  local reached = resolved(self.base_url, path, query)
  local deadline = socket.gettime() + self.timeout

  local written = nil
  if body ~= nil then
    local ok, encoded_body = Json.try_encode(body)
    if not ok then
      -- A caller's mistake rather than a failure of the exchange: nothing about the network would
      -- make the same value encodable the second time, and a send has nothing to decide here.
      Errors.throw(Errors.ClientError, "the body of the request is not something a document can carry")
    end
    written = encoded_body
  end

  local head = {
    string.upper(tostring(method)) .. " " .. reached.target .. " HTTP/1.1",
    "Host: " .. reached.host .. (reached.port == PORTS[reached.scheme] and "" or ":" .. reached.port),
    "Authorization: Bearer " .. tostring(self.token),
    "Accept: " .. Transport.JSON_MEDIA_TYPE,
    "User-Agent: " .. user_agent(),
    "Hook0-Client-Options: " .. client_options(self.retry_policy),
    "Connection: close",
  }
  if written ~= nil then
    head[#head + 1] = "Content-Type: " .. Transport.JSON_MEDIA_TYPE
    head[#head + 1] = "Content-Length: " .. #written
  end

  local connection = connected(reached, deadline)
  local ok, answered = pcall(function()
    local sent, failed = within(connection, deadline):send(
      table.concat(head, "\r\n") .. "\r\n\r\n" .. (written or ""))
    if sent == nil then
      Transport.no_answer("the request could not be sent: " .. tostring(failed))
    end

    local status = head_line(connection, deadline, self.max_header_bytes):match("^HTTP/%d%.%d%s+(%d%d%d)")
    if status == nil then
      Transport.no_answer("the API answered something that is not an HTTP status line")
    end
    local headers = carried(connection, deadline, self)
    return {
      math.tointeger(tonumber(status)),
      headers,
      bounded(connection, deadline, headers, self.max_response_bytes),
    }
  end)

  connection:close()
  if not ok then
    error(answered, 0)
  end
  return answered[1], answered[2], answered[3]
end

--- What the API answered, whether or not it answered a success.
---
--- This is the shape the generated half of this rock reads, which is the status and the bytes. A
--- caller that also needs what the answer carried beside its body — the delay a paced instance names
--- is one — asks `deliver` for it.
---
--- @param method string the HTTP method the operation is issued under
--- @param path string where the request lands, absolute or under the base URL
--- @param query table|nil the name and value pairs of the query string
--- @param body any|nil what to send as a JSON document, or nothing at all
--- @return integer, string the status and the body
function Transport:request(method, path, query, body)
  local status, _, payload = self:deliver(method, path, query, body)
  return status, payload
end

return Transport
